File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/view/ScrollLayoutView.java
/* Copyright Airship and Contributors */

package com.urbanairship.android.layout.view;

import android.content.Context;
import android.view.View;

import com.urbanairship.android.layout.Thomas;
import com.urbanairship.android.layout.environment.ViewEnvironment;
import com.urbanairship.android.layout.model.ScrollLayoutModel;
import com.urbanairship.android.layout.property.Direction;
import com.urbanairship.android.layout.util.LayoutUtils;

import androidx.annotation.NonNull;
import androidx.core.view.ViewCompat;
import androidx.core.widget.NestedScrollView;

public class ScrollLayoutView extends NestedScrollView implements BaseView {
    private final ScrollLayoutModel model;
    private final ViewEnvironment viewEnvironment;

    public ScrollLayoutView(
            @NonNull Context context,
            @NonNull ScrollLayoutModel model,
            @NonNull ViewEnvironment viewEnvironment
    ) {
        super(context);
        this.model = model;
        this.viewEnvironment = viewEnvironment;

        setId(model.getViewId());

        configure();
    }

    private void configure() {
        setFillViewport(false);
        LayoutUtils.applyBorderAndBackground(this, model);

        Direction direction = model.getDirection();
        View contentView = Thomas.view(getContext(), model.getView(), viewEnvironment);

        LayoutParams layoutParams;
        if (direction == Direction.VERTICAL) {
            layoutParams = new LayoutParams(LayoutParams.MATCH_PARENT, LayoutParams.WRAP_CONTENT);
        } else {
            layoutParams = new LayoutParams(LayoutParams.WRAP_CONTENT, LayoutParams.MATCH_PARENT);
        }

        setClipToOutline(true);
        contentView.setLayoutParams(layoutParams);
        addView(contentView);

        // Pass along any calls to apply insets to the view.
        ViewCompat.setOnApplyWindowInsetsListener(this, (v, insets) ->
            ViewCompat.dispatchApplyWindowInsets(contentView, insets)
        );
    }
}

File: urbanairship-layout/src/main/java/com/urbanairship/android/layout/view/ScrollLayoutView.kt
/* Copyright Airship and Contributors */
package com.urbanairship.android.layout.view

import android.content.Context
import android.view.View
import android.widget.FrameLayout.LayoutParams.MATCH_PARENT
import android.widget.FrameLayout.LayoutParams.WRAP_CONTENT
import androidx.core.view.ViewCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.isGone
import androidx.core.view.isVisible
import androidx.core.widget.NestedScrollView
import com.urbanairship.android.layout.environment.ViewEnvironment
import com.urbanairship.android.layout.model.Background
import com.urbanairship.android.layout.model.BaseModel
import com.urbanairship.android.layout.model.ScrollLayoutModel
import com.urbanairship.android.layout.property.Border
import com.urbanairship.android.layout.property.Color
import com.urbanairship.android.layout.property.Direction
import com.urbanairship.android.layout.util.LayoutUtils

internal class ScrollLayoutView(
    context: Context,
    model: ScrollLayoutModel,
    viewEnvironment: ViewEnvironment
) : NestedScrollView(context), BaseView {

    init {
        isFillViewport = false
        clipToOutline = true

        val contentView = model.view.createView(context, viewEnvironment, null).apply {
            layoutParams = if (model.viewInfo.direction == Direction.VERTICAL) {
                LayoutParams(MATCH_PARENT, WRAP_CONTENT)
            } else {
                LayoutParams(WRAP_CONTENT, MATCH_PARENT)
            }
        }
        addView(contentView)

        model.listener = object : BaseModel.Listener {
            override fun setVisibility(visible: Boolean) {
                this@ScrollLayoutView.isVisible = visible
            }
            override fun setEnabled(enabled: Boolean) {
                this@ScrollLayoutView.isEnabled = enabled
            }

            override fun setBackground(old: Background?, new: Background) {
                LayoutUtils.updateBackground(this@ScrollLayoutView, old, new)
            }
        }

        // Pass along any calls to apply insets to the view.
        ViewCompat.setOnApplyWindowInsetsListener(this) { _: View, insets: WindowInsetsCompat ->
            ViewCompat.dispatchApplyWindowInsets(contentView, insets)
        }
    }
}
